### PR TITLE
ACL implementation in create-service work flow

### PR DESCRIPTION
**Pull request by supritat**
### Requirements

* ACL implementation: Update admin permission while creating service.

### Description of the Change

* ACL implementation in create-serverless-service API to provide admin permission for manage category
*  ACL implementation in service-onboarding-build-pack jenkins file to provide admin permission for code and deploy category

### Benefits

Service owner will be getting admin permission for all the categories.

### Possible Drawbacks

NA
### Applicable Issues

NA

### CODE REVIEW

File: builds/service-onboarding-build-pack/Jenkinsfile [hunk 1]
```diff
@@ -452,3 +452,38 @@ def jsonParse(jsonString) {
...
 }
+
+def updateAclPermission(serviceId, userId) {
+	try {
+		def categoryList = ['code', 'deploy']
+		def policiesList = []
+		for (category in categoryList) {
+			def eachPolicy = [
+				userId: userId,
+				permission: 'admin',
```
Comment (svidhani) [inline builds/service-onboarding-build-pack/Jenkinsfile]: for code and deploy its not admin but write
Manage - read|write|admin
Code - read | write
Deploy - read|write

File: core/jazz_create-serverless-service/index.js [hunk 2]
```diff
@@ -357,10 +359,49 @@ var validateEventName = (eventType, sourceName, config) => {
...
+                        }
+                    ]
+                }
+            ],
+            rejectUnauthorized: false
+        };
+
+        request(svcPayload, (error, response, body) => {
+            if(error) {
+                logger.error(`Error while updating policies t ACL: ${error}`);
```
[the remark this answers is not in the document]
Comment (supritat) [reply core/jazz_create-serverless-service/index.js]: updated

File: core/jazz_create-serverless-service/index.js [hunk 3]
```diff
@@ -357,10 +359,49 @@ var validateEventName = (eventType, sourceName, config) => {
...
+                reject(error);
+            } else {
+                logger.debug(`ACL response: ${JSON.stringify(response)}`);
+                if(body && body.data && body.data.success) {
+                    resolve();
+                } else {
+                    logger.error(`Error while updating policies t ACL: ${JSON.stringify(response)}`);
```
[the remark this answers is not in the document]
Comment (supritat) [reply core/jazz_create-serverless-service/index.js]: updated

File: builds/service-onboarding-build-pack/Jenkinsfile (hunk 1 above)
Comment (supritat) [reply builds/service-onboarding-build-pack/Jenkinsfile]: Updated

File: builds/service-onboarding-build-pack/Jenkinsfile [hunk 4]
```diff
@@ -452,3 +452,38 @@ def jsonParse(jsonString) {
     def nonLazyMap = new groovy.json.JsonSlurperClassic().parseText(jsonString)
     return nonLazyMap
 }
+
+def updateAclPermission(serviceId, userId) {
```
Comment (SiniWilson) [inline builds/service-onboarding-build-pack/Jenkinsfile]: Please move this code to a separate module.

File: builds/service-onboarding-build-pack/Jenkinsfile [hunk 5]
```diff
@@ -452,3 +452,38 @@ def jsonParse(jsonString) {
...
+			policies: policiesList
+		]);
+		def updatePermission = sh(script: "curl -X POST \
+				${g_base_url}/jazz/acl/policies \
+				-k -v -H \"Authorization: $auth_token\" \
+				-H \"Content-Type: application/json\" \
+				-d \'${body}\'", returnStdout: true).trim()
+		def responseJSON = parseJson(updatePermission)
+
+		if (responseJSON && responseJSON.success) {
```
Comment (SiniWilson) [inline builds/service-onboarding-build-pack/Jenkinsfile]: Make sure the response contains "**success**"

File: core/jazz_create-serverless-service/index.js [hunk 6]
```diff
@@ -357,10 +359,49 @@ var validateEventName = (eventType, sourceName, config) => {
...
+            rejectUnauthorized: false
+        };
+
+        request(svcPayload, (error, response, body) => {
+            if(error) {
+                logger.error(`Error while updating policies using ACL: ${error}`);
+                reject(error);
+            } else {
+                logger.debug(`ACL response: ${JSON.stringify(response)}`);
+                if(body && body.data && body.data.success) {
```
Comment (SiniWilson) [inline core/jazz_create-serverless-service/index.js]: Make sure the response contains "success"

File: core/jazz_create-serverless-service/index.js [hunk 7]
```diff
@@ -357,10 +359,49 @@ var validateEventName = (eventType, sourceName, config) => {
...
+                    resolve();
+                } else {
+                    logger.error(`Error while updating policies using ACL: ${JSON.stringify(response)}`);
+                    reject(response);
```
Comment (SiniWilson) [inline core/jazz_create-serverless-service/index.js]: why rejecting "response" ?

File: builds/service-onboarding-build-pack/Jenkinsfile (hunk 4 above)
Comment (supritat) [reply builds/service-onboarding-build-pack/Jenkinsfile]: moved to scm-module

File: builds/service-onboarding-build-pack/Jenkinsfile (hunk 5 above)
Comment (supritat) [reply builds/service-onboarding-build-pack/Jenkinsfile]: Done

File: core/jazz_create-serverless-service/index.js (hunk 6 above)
Comment (supritat) [reply core/jazz_create-serverless-service/index.js]: Updated

File: core/jazz_create-serverless-service/index.js (hunk 7 above)
Comment (supritat) [reply core/jazz_create-serverless-service/index.js]: updated.

File: builds/jazz-build-module/scm-module.groovy [hunk 8]
```diff
@@ -338,4 +338,27 @@ def getRepoCommitterInfo(commitHash) {
     }
 }
 
+def updateAclPermission(serviceId, policiesList, auth_token, aclUrl) {
```
Comment (devsatishm) [inline builds/jazz-build-module/scm-module.groovy]: As discussed in https://github.com/tmobile/jazz/pull/676, this has to move out of SCM module.

Comment (supritat) [reply builds/jazz-build-module/scm-module.groovy]: moved to acl-module

File: builds/service-onboarding-build-pack/Jenkinsfile [hunk 9]
```diff
@@ -245,7 +245,19 @@ node  {
 
 				try {
 					events.sendStartedEvent("ADD_WRITE_PERMISSIONS_TO_SERVICE_REPO", 'adding write permission to repo started', context_map)
-					scmModule.setRepoPermissions(service_config['created_by'], repo_name, admin_group)
+					def aclUrl = "${g_base_url}/jazz/acl/policies"
```
[the remark this answers is not in the document]
Comment (supritat) [reply builds/service-onboarding-build-pack/Jenkinsfile]: Updated.

File: core/jazz_create-serverless-service/test/test.js [hunk 10]
```diff
@@ -722,4 +727,44 @@ describe('create-serverless-service', function () {
...
+      let error = {errType: "invalidError", errMessage: "error occoured"}
+      let reqStub = sinon.stub(request, "Request", (obj) => {
+        return obj.callback(error, null, null);
+      });
+      index.updateAclPolicy("random-id", 'temp-token', 'user', 'temp-permission', 'temp-cat', config)
+      .catch(err => {
+        expect(err).to.include(error);
+        sinon.assert.calledOnce(reqStub);
+        reqStub.restore();
+      })
```
[the remark this answers is not in the document]
Comment (supritat) [reply core/jazz_create-serverless-service/test/test.js]: Worked on this.

**Pull request merged**